src/x/lockfile/lockfile.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package lockfile
+
+import (
+	"os"
+	"syscall"
+
+	"github.com/pkg/errors"
+)
+
+type Lockfile struct {
+	file os.File
+}
+
+// Create creates the given file path if it doesn't exist and
+// obtains an exclusive lock on it. An error is returned if the lock
+// has been obtained by another process.
+func Create(path string) (*Lockfile, error) {
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, 0666)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed opening lock path")
+	}
+
+	ft := &syscall.Flock_t{
+		Pid:  int32(os.Getpid()),
+		Type: syscall.F_WRLCK,
+	}
+
+	if err = syscall.FcntlFlock(file.Fd(), syscall.F_SETLK, ft); err != nil {
+		return nil, errors.Wrap(err, "failed obtaining lock")
+	}
+
+	lf := Lockfile{*file}
+
+	return &lf, nil
+}
+
+// Remove releases the lock on the file and removes the file.
+func (lf Lockfile) Remove() error {
+	ft := &syscall.Flock_t{
+		Pid:  int32(os.Getpid()),
+		Type: syscall.F_UNLCK,
+	}
+
+	if err := syscall.FcntlFlock(lf.file.Fd(), syscall.F_SETLK, ft); err != nil {
+		return errors.Wrap(err, "failed releasing lock")
+	}
+
+	if err := os.Remove(lf.file.Name()); err != nil {
```

Comment:
is it idiomatic to remove and then close?

Reply:
I did some quick research but I didn't see any example of first remove then close -- I modified the code to close, then remove. I was thinking that there might be a race condition:
- process A closes the file handle
- process B creates the file and obtains the lock
- process A remove the files
- at this point B has a lock on a file which has been removed from the file system

But as I said, the code now closes, then removes, the above scenario doesn't seem likely. Let me know if you think we should dig deeper here and I'll try to reproduce the above scenario to check if it's a real concern.

Reply:
@edkolev If you delete a file that another process has a handle on, the file won't actually be deleted until that process releases the file